README.md
# Floating Text

<p>
<a href="https://flutter.dev">
    <img src="https://img.shields.io/badge/Platform-Flutter-02569B?logo=flutter" alt="Platform">
  </a>
  <a href="https://pub.dev/packages/floating_text" rel="ugc">
    <img src="https://img.shields.io/badge/pub-0.2.0-blue" alt="Pub Package">
  </a>
  <a href="https://opensource.org/licenses/MIT" rel="ugc">
    <img src="https://img.shields.io/badge/likes-26-green" alt="License: MIT">
  </a>
  <a href="https://opensource.org/licenses/MIT" rel="ugc">
    <img src="https://img.shields.io/badge/license-MIT-red" alt="License: MIT">
  </a>
</p>

A light weight flutter package that provides beautiful floating text animations.

## Getting Started

Add this package in your project's dependencies.

```yaml
dependencies:
  floating_text: ^0.2.0
```

Import the package:

```dart
import 'package:floating_text/floating_text.dart'
```

Use Anywhere instead of `Text` Widget:

## Demo's

<table>
  <tr>
    <th>Demo 1</th>
    <th>Demo 2</th>
    <th>Demo 3</th>
  </tr>
  <tr>
    <td><img alt = 'floating_text down float animation' src = "https://raw.githubusercontent.com/pktintali/floating_text/main/example/screenshots/floating_down.gif" /></td>
    <td><img alt = 'animation with shadow' src = "https://raw.githubusercontent.com/pktintali/floating_text/main/example/screenshots/animation_with_shadow.gif" /></td>
    <td><img alt = 'linear animation' src = "https://raw.githubusercontent.com/pktintali/floating_text/main/example/screenshots/floating_linear.gif" /></td>
  </tr>
</table>
<table>
  <tr>
    <th>Demo 4</th>
    <th>Demo 5</th>
    <!-- <th>Demo 6</th> -->
  </tr>
  <tr>
    <td><img alt = 'animation on button' src = "https://raw.githubusercontent.com/pktintali/floating_text/main/example/screenshots/button_text.gif" /></td>
    <td><img alt = 'floating_text up float animation' src = "https://raw.githubusercontent.com/pktintali/floating_text/main/example/screenshots/floating_up.gif" /></td>
    <!-- <td><img alt = 'rtl support' src = "https://raw.githubusercontent.com/pktintali/floating_text/main/example/screenshots/rtl.gif" /></td> -->
  </tr>
</table>

## Key Features

- Light Weight ( Only 100 Lines of code inside package )
- Support any Text on the Screen
- Custom style for the floating/animated text
- Custom style support for the non floating text
- highly customizable
- Easy to use

### Use On Any Widget

```dart
RaisedButton(
  onPressed: () {},
  child: FloatingText(
    text: 'Button',
    repeat: true,
    duration: Duration(milliseconds: 500),
  ),
),
```

<!-- ### RTL Support

```dart
FloatingText(
   text: 'WELCOME',
   repeat: true,
   isRTL: true,
   duration: Duration(milliseconds: 100),
 ),
``` -->

### Set repeat true for repeating the animation

```dart
FloatingText(
text: "WELCOME",
repeat: true,
duration: Duration(milliseconds: 600),
floatingTextStyle: TextStyle(
  color: Colors.blue,
  fontSize: 50,
),
textStyle: TextStyle(
color: Colors.black38,
  fontSize: 60,
)),
onAnimationComplete: () {
  //Do your stuff
},
```

### Use your custom text style

```dart
FloatingText(
    text: 'Congratulations',
    repeat: true,
    textStyle: TextStyle(
      fontSize: 40,
      color: Colors.black54,
    ),
    floatingTextStyle: TextStyle(
      color: Colors.red,
      fontSize: 40,
      shadows: [
        BoxShadow(
          color: Colors.yellow,
          blurRadius: 10,
        )
      ],
    ),
  ),
```

### Full Example

```dart
FloatingText(
    text: 'Congratulations',
    repeat: true,
    repeatCount: 5,
    isRTL: false,
    duration: Duration(milliseconds: 100),
    textStyle: TextStyle(
      fontSize: 40,
      color: Colors.black54,
    ),
    floatingTextStyle: TextStyle(
      color: Colors.red,
      fontSize: 40,
      shadows: [
        BoxShadow(
          color: Colors.yellow,
          blurRadius: 10,
        )
      ],
    ),
    onAnimationComplete: () {
      print('Animation Completed');
    },
  ),
```

## Parameters

- `text` (String) - text to display on screen
- `repeat` (Boolean) - by default `false`, set `true` if you want animation to repeat
- `repeatCount` (Integer) - Number of time animation should repeat. _repeat_ paramater must be `true` for it to work.
- `duration` (Duration) - duration of the animation. Default value 200 milliseconds
- `isRTL` (Boolean) - by default `false`, set `true` for RTL support
- `textStyle` (TextStyle) - text style for floating/animating part of text
- `floatingTextStyle` (TextStyle) - text style for not floating part of text
- `onAnimationComplete` (Function - void) - Called once the animation is complete.

> Tip: Experiment with different `textStyle` and non `floatingTextStyle` to get some new cool effects


CHANGELOG.md
# Changelog

## [0.2.0] - 01/09/2022

### What's new

- Callback `onAnimationComplete` added. It will be called when the animation is completed.
- Property `repeatCount` added. It defines how many time the animation should repeat. `repeat` must be _true_ for it to work.

### What's fixed

- TextOverflow problem fixed.
- RTL support fixed.

## [0.1.1] - 13/09/2021

- readme updated

## [0.1.0] - 12/06/2021

- null safety in stable

## [0.0.4] - 12/02/2021

- floatingSpeed parameter changed to duration

## [0.0.3] - 28/12/2020

- RTL Support Updated

## [0.0.2] - 23/12/2020

- Updated Readme and documentation typo fixed.

## [0.0.1] - 22/12/2020

- Initial release, New Animations and Features Coming Soon.


lib/src/FloatingText.dart
import 'dart:async';
import 'package:flutter/material.dart';

class FloatingText extends StatefulWidget {
  /// Text for [Text] widget.
  final String text;

  ///Should [Animation] repeat
  ///
  /// boolean, by default set to false
  final bool repeat;

  ///Integer, Number of times animation should repeat
  ///
  ///If not given, animation will not stop
  ///
  ///repeat should be true to work
  final int? repeatCount;

  ///What shoult happend once animation completed
  final VoidCallback? onAnimationComplete;

  ///Animation speed for the [FloatingText]
  ///
  /// By Default Set to 200 milliseconds
  final Duration duration;

  ///[TextStyle] for non floating [Text]
  final TextStyle textStyle;

  ///[TextStyle] for floating [Text]
  final TextStyle floatingTextStyle;

  ///Enable RTL support
  ///
  ///boolean by default set to false
  final bool isRTL;

  ///key
  final Key? key;
  const FloatingText(
      {required this.text,
      this.key,
      this.duration = const Duration(milliseconds: 200),
      this.isRTL = false,
      this.textStyle = const TextStyle(),
      this.floatingTextStyle = const TextStyle(color: Colors.red),
      this.repeat = false,
      this.onAnimationComplete,
      this.repeatCount});
  @override
  _FloatingTextState createState() => _FloatingTextState();
}

class _FloatingTextState extends State<FloatingText> {
  List<String> _sList = [];
  List<Widget> _tList = [];
  String? _pString;
  late Timer _timer;
  double _temp = 0;
  Duration? dur;
  int? _loopCount;
  void changePosition(Timer t) async {
    if ((widget.repeat && _temp != -1) || _temp < widget.text.length + 1) {
      _tList.clear();
      for (int i = 0; i < _sList.length; i++) {
        _tList.add(
          Text(
            _sList[i],
            style: i == _temp ? widget.floatingTextStyle : widget.textStyle,
          ),
        );
      }
      if (_temp == widget.text.length + 1) {
        _temp = -1;
        if (_loopCount != null) {
          _loopCount = _loopCount! - 1;
        }
        if (_loopCount == 0) {
          _timer.cancel();
          widget.onAnimationComplete!();
        }
      }
      setState(() {
        _temp++;
      });
    } else {
      _timer.cancel();
      widget.onAnimationComplete!();
    }
  }

  void _splitString() {
    _sList.clear();
    String s = widget.text;
    for (int i = 0; i < widget.text.length; i++) {
      _sList.add(s[i]);
    }
    _temp = 0;
  }

  void _setTimer() {
    _timer.cancel();
    _timer = Timer.periodic(dur!, changePosition);
  }

  @override
  void initState() {
    _loopCount = widget.repeatCount;
    _splitString();
    _temp = 0;
    _timer = Timer.periodic(widget.duration, changePosition);
    super.initState();
  }

  @override
  void dispose() {
    _timer.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (widget.text != _pString) {
      _splitString();
      _pString = widget.text;
    }

    if (widget.duration != dur) {
      dur = widget.duration;
      _setTimer();
    }

    return Wrap(
      textDirection: widget.isRTL ? TextDirection.rtl : TextDirection.ltr,
      key: widget.key,
      crossAxisAlignment: WrapCrossAlignment.center,
      children: widget.isRTL
          ? _tList.reversed.map((s) {
              return s;
            }).toList()
          : _tList.map((s) {
              return s;
            }).toList(),
    );
  }
}


example/lib/main.dart
//Author : Pradeep Tintali
//github : github.com/pktintali

import 'package:flutter/material.dart';
import 'package:floating_text/floating_text.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'FloatingText Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: CustomCardsExample(),
    );
  }
}

class CustomCardsExample extends StatefulWidget {
  @override
  _CustomCardsExampleState createState() => _CustomCardsExampleState();
}

class _CustomCardsExampleState extends State<CustomCardsExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: const Text(
          'FloatingText Example',
        ),
      ),
      body: Center(
        //Using Floating Text as a Widget
        child: FloatingText(
          onAnimationComplete: () {
            print('Animation Completed');
          },
          repeat: true,
          repeatCount: 3,
          text: 'Congratulations',
          duration: Duration(milliseconds: 150),
          //Customize your text
          textStyle: const TextStyle(
            fontSize: 40,
            color: Colors.black54,
          ),
          //Customize your Floating text
          floatingTextStyle: const TextStyle(
            color: Colors.red,
            fontSize: 50,
            shadows: const [
              BoxShadow(
                color: Colors.yellow,
                blurRadius: 10,
              )
            ],
          ),
        ),
      ),
    );
  }
}
